### ISSUE TITLE
Parsing Failed

### ISSUE THREAD

**Issue by akkinenirajesh** [javaparser/javaparser#149]
Failed to parse any java file with typed lambda params. (code copied from java.time.chrono.AbstractChronology.java)

```
static final Comparator<ChronoLocalDate> DATE_ORDER =
    (Comparator<ChronoLocalDate> & Serializable) (date1, date2) -> {
        return Long.compare(date1.toEpochDay(), date2.toEpochDay());
    };
```

**Comment by ftomassetti** [comment]
The problem is with the "combined cast": if you remove & Serializable it can be parsed. I started writing a test case and looking into the grammar for a solution.

**Comment by akkinenirajesh** [comment]
Yes, I noticed. I am new to JavaCC so taking time for me to understand where to change. It seems to a new kind of Casting 
    static final Consumer<String> x=(Consumer<String> & Serializable)null; 
this seems to be valid in Java 8. Just tested in Eclipse. So changes should be in Casting Expression I believe.

**Comment by ftomassetti** [comment]
Yes, I also think so. In my opinion the relevant piece of the grammar is:

```
Expression CastExpression():
{
    Expression ret;
    Type type;
    int line;
    int column;
    AnnotationExpr ann;
    List annotations = null;
}
{
  "(" {line=token.beginLine; column=token.beginColumn;}
   (ann = Annotation()   { annotations = add(annotations, ann);})*
  (
      LOOKAHEAD(2)
      type = PrimitiveType() ")" ret = UnaryExpression() { type.setAnnotations(annotations); ret = new CastExpr(line, column, token.endLine, token.endColumn,type, ret); }
    |
      type = ReferenceType() ")" ret = UnaryExpressionNotPlusMinus() { type.setAnnotations(annotations); ret = new CastExpr(line, column, token.endLine, token.endColumn,type, ret); }
  )
  { return ret; }
}
```

The problem is how to find a sensible representation of this new cast without changing the metamodel of the currently existing CastExpression.

**Comment by akkinenirajesh** [comment]
How does "create a new CombinedCastExpr that will contain one or more CastExpr " sound ?
In this way we don't have to break any existing code.

**Comment by akkinenirajesh** [comment]
Sorry, I think my idea is stupid. I did not think through...

**Comment by ftomassetti** [comment]
With this change it parses the example given:

```
Expression CastExpression():
{
    Expression ret;
    Type type;
    int line;
    int column;
    AnnotationExpr ann;
    Type st;
    List annotations = null;
    List secondaryTypes = null;
}
{
  "(" {line=token.beginLine; column=token.beginColumn;}
   (ann = Annotation()   { annotations = add(annotations, ann);})*
  (
      LOOKAHEAD(2)
      type = PrimitiveType() ")" ret = UnaryExpression() { type.setAnnotations(annotations); ret = new CastExpr(line, column, token.endLine, token.endColumn,type, ret); }
    |
      type = ReferenceType() ("&" st = ReferenceType() { secondaryTypes = add(secondaryTypes, st);} )* ")" ret = UnaryExpressionNotPlusMinus() { type.setAnnotations(annotations); ret = new CastExpr(line, column, token.endLine, token.endColumn,type, ret); }
  )
  { return ret; }
}
```

but it introduces a new field called "secondaryTypes" in the cast expression. The problem is that the cast expression is quite subtle because it is very ambigous and so it has to be parsed using lookahead:

For example `(foo)` could be a cast or a reference to a field between parenthesis.

I think it would make difficult to introduce a new CombinedCast expression. In addition to that I guess there is code out there looking for CastExpr but not expecting CombinedCastExpr. I guess there are different possible solutions with pros and cons. For now I will open a pull request and ask for comments from you and other contributors.

And thank you for reporting this one!

**Comment by akkinenirajesh** [comment]
How about adding a MultiType extends Type? is it too much? this will limit the changes.
We can make MultiType behave as the first type(not sure if that is a problem) in it.

**Comment by ftomassetti** [comment]
This is also an interesting idea (in the meantime the PR is available at #150). Do you mind suggest this also there?

**Comment by ftomassetti** [comment]
We closed the previous PR because it is probably better to go for a clean solution breaking compatibility

**Comment by ftomassetti** [comment]
A new PR has been presented

**Comment by matozoid** [comment]
What is the status of this?

**Comment by ftomassetti** [comment]
I think it was solved through #185

**Comment by matozoid** [comment]
Yes, I spotted some tests for it in that PR.
